include/pisa/query/algorithm/range_query.hpp

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include "topk_queue.hpp"
+#include "query/queries.hpp"
+
+namespace pisa {
+
+template<typename QueryAlg>
+struct range_query {
+
+    range_query(uint64_t k, uint64_t max_docid, uint64_t range_size)
+        : m_k(k), m_topk(k), m_max_docid(max_docid), m_range_size(range_size) {}
+
+    template<typename CursorRange>
+    uint64_t operator()(CursorRange &&cursors) {
+        m_topk.clear();
+        if (cursors.empty()) {
+            return 0;
+        }
+
+        size_t end = m_range_size;
+        for (; end <= m_max_docid; end += m_range_size) {
+            processRange(cursors, end);
+        }
+        if (end < m_max_docid + m_range_size) {
```

Comment:
Tip: `if` statements are bad for efficiency! if you can avoid them, you should do it.
If you want to know more, look for `if branch misprediction`.

That said, here we can avoid this check.
```cpp
// Loop excluding always the very last range 
for (size_t end = m_range_size; end + m_range_size <= m_max_docid; end += m_range_size) {
   processRange(cursors, end);
}
// Always process the last one
processRange(cursors, m_max_docid);
```

Reply:
Not sure if that's gonna have much impact here, but Antonio's way is definitely cleaner. And a good tip in general.